### PR TITLE
Start dropping spans when the span exporter thread holds more than 10K spans references.

### PR DESCRIPTION
**Pull request by bogdandrutu**
Fixes https://github.com/census-instrumentation/opencensus-java/issues/1767

### CODE REVIEW

File: impl_core/src/main/java/io/opencensus/implcore/trace/export/SpanExporterImpl.java [hunk 1]
```diff
@@ -154,6 +254,7 @@ private void onBatchExport(List<SpanData> spanDataList) {
     private Worker(int bufferSize, Duration scheduleDelay) {
       spans = new ArrayList<>(bufferSize);
       this.bufferSize = bufferSize;
+      this.maxReferencedSpans = 10L * bufferSize;
```
Comment (songy23) [inline impl_core/src/main/java/io/opencensus/implcore/trace/export/SpanExporterImpl.java]: Any specific reason we chose this value? Do we want to make it configurable?

Comment (rghetia) [reply impl_core/src/main/java/io/opencensus/implcore/trace/export/SpanExporterImpl.java]: I was wondering the same.

File: impl_core/src/main/java/io/opencensus/implcore/trace/export/SpanExporterImpl.java [hunk 2]
```diff
@@ -51,7 +85,7 @@
    */
   static SpanExporterImpl create(int bufferSize, Duration scheduleDelay) {
     // TODO(bdrutu): Consider to add a shutdown hook to not avoid dropping data.
-    Worker worker = new Worker(bufferSize, scheduleDelay);
+    final Worker worker = new Worker(bufferSize, scheduleDelay);
```
Comment (dinooliva) [inline impl_core/src/main/java/io/opencensus/implcore/trace/export/SpanExporterImpl.java]: Nit: why 'final'?

File: impl_core/src/main/java/io/opencensus/implcore/trace/export/SpanExporterImpl.java [hunk 3]
```diff
@@ -32,9 +38,37 @@
...
 
+/*>>>
+import org.checkerframework.checker.nullness.qual.Nullable;
+*/
+
 /** Implementation of the {@link SpanExporter}. */
 public final class SpanExporterImpl extends SpanExporter {
   private static final Logger logger = Logger.getLogger(ExportComponent.class.getName());
+  private static final DerivedLongCumulative droppedSpans =
+      Metrics.getMetricRegistry()
```
Comment (dinooliva) [inline impl_core/src/main/java/io/opencensus/implcore/trace/export/SpanExporterImpl.java]: Tracing impl now depends on Metrics - that seems fine - I prefer not to have cyclic dependencies but, iiuc, this doesn't introduce them. Interesting issue, tho - if we split the artifacts for tracing and metrics, this won't work.

Comment (bogdandrutu) [reply impl_core/src/main/java/io/opencensus/implcore/trace/export/SpanExporterImpl.java]: It depends only on the API artifact not the impl artifact.

Comment (dinooliva) [reply impl_core/src/main/java/io/opencensus/implcore/trace/export/SpanExporterImpl.java]: My second point was that a service that links in a tracing impl but not a metrics impl will get unexpected results - not an issue with the current way we package artifacts but could be a problem if we ever decide to split the implementation artifacts.

File: impl_core/src/main/java/io/opencensus/implcore/trace/export/SpanExporterImpl.java (hunk 2 above)
Comment (bogdandrutu) [reply impl_core/src/main/java/io/opencensus/implcore/trace/export/SpanExporterImpl.java]: Done.

File: impl_core/src/main/java/io/opencensus/implcore/trace/export/SpanExporterImpl.java [hunk 4]
```diff
@@ -110,14 +195,29 @@ Thread getServiceExporterThread() {
...
+
+    private final Map<String, Handler> serviceHandlers = new ConcurrentHashMap<>();
     private final int bufferSize;
+    private final long maxReferencedSpans;
     private final long scheduleDelayMillis;
 
     // See SpanExporterImpl#addSpan.
     private void addSpan(RecordEventsSpanImpl span) {
       synchronized (monitor) {
+        if (referencedSpans + 1L > maxReferencedSpans) {
```
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply impl_core/src/main/java/io/opencensus/implcore/trace/export/SpanExporterImpl.java]: done.

File: impl_core/src/main/java/io/opencensus/implcore/trace/export/SpanExporterImpl.java (hunk 1 above)
Comment (bogdandrutu) [reply impl_core/src/main/java/io/opencensus/implcore/trace/export/SpanExporterImpl.java]: Added some explanation about why I decided to go with 4 * bufferSize in the end.

**Pull request merged**